### PR TITLE
Create login functionality

### PR DESCRIPTION
**Pull request by allthesignals**
This commit creates a sign-in component, which encapsulates all the login behavior. Tests are created for this, including acceptance and integration.

A new package, window mock, is included to assist with testing behavior that depends on redirects.

# Conflicts:
#	client/app/authenticators/zap-api-authenticator.js
#	client/app/templates/application.hbs

### CODE REVIEW

File: client/app/models/user.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import Model from '@ember-data/model';
+
+export default class UserModel extends Model {
+
```
Comment (godfreyyeung) [inline client/app/models/user.js:4]: Will need attributes:
- contactId <-- camelcased here, I think
- emailaddress1

File: client/app/router.js [hunk 2]
```diff
@@ -9,4 +9,5 @@ export default class Router extends EmberRouter {
 
 Router.map(function() { // eslint-disable-line
   this.route('projects');
```
Comment (godfreyyeung) [inline client/app/router.js:11]: wondering if we can nest authenticated routes under 
```
  this.route('authenticated', { path: '' }, function() {
```

File: client/app/models/user.js (hunk 1 above)
Comment (allthesignals) [reply client/app/models/user.js:4]: yup! not camelcase tho

File: client/app/router.js (hunk 2 above)
Comment (allthesignals) [reply client/app/router.js:11]: great idea. enhancement for later?

File: client/app/templates/application.hbs [hunk 3]
```diff
@@ -5,15 +5,26 @@
...
+      <li>
+        {{#if this.session.isAuthenticated}}
+          <button {{on "click" this.invalidateSession}}
+            href="#"
+            type="button"
+          >
+            Logout
+          </button>
+        {{else}}
+          <SignInButton />
```
Comment (bfreeds) [inline client/app/templates/application.hbs:25]: @allthesignals I think the layout in the sign-in button component needs to be adjusted:
image

Comment (allthesignals) [reply client/app/templates/application.hbs:25]: bah crap. it needs to be a button element bc of accessibbility. can't be anchor.

Comment (allthesignals) [reply client/app/templates/application.hbs:25]: @andycochran need some help with this — this is styled for anchor tags, but we need to use button elements bc you can't bind actions to anchor elements. thoughts?

I'm going to move into new issue so we can move on

**Pull request merged**